engine/src/main/scala/cromwell/engine/workflow/tokens/TokenPool.scala

```diff
@@ -0,0 +1,49 @@
+package cromwell.engine.workflow.tokens
+
+import java.util.UUID
+
+import cromwell.core.JobExecutionToken
+import JobExecutionToken.JobExecutionTokenType
+import cromwell.engine.workflow.tokens.TokenPool.TokenPoolPop
+
+import scala.language.postfixOps
+
+trait TokenPool {
+  def currentLoans: Set[JobExecutionToken]
+  def pop(): TokenPoolPop
+  def push(jobExecutionToken: JobExecutionToken): TokenPool
+}
+
+object TokenPool {
+
+  case class TokenPoolPop(newTokenPool: TokenPool, poppedItem: Option[JobExecutionToken])
+
+  def apply(tokenType: JobExecutionTokenType): TokenPool = {
+    tokenType.maxPoolSize map { ps =>
+      val pool = (1 to ps toList) map { _ => JobExecutionToken(tokenType, UUID.randomUUID()) }
+      SizedTokenPool(pool.toSet, Set.empty)
+    } getOrElse {
+      InfiniteTokenPool(tokenType, Set.empty)
+    }
+  }
+}
+
+final case class SizedTokenPool(pool: Set[JobExecutionToken], override val currentLoans: Set[JobExecutionToken]) extends TokenPool {
+
+  override def pop(): TokenPoolPop = pool.toList match {
+    case head :: tail => TokenPoolPop(SizedTokenPool(tail.toSet, currentLoans + head), Some(head))
+    case Nil => TokenPoolPop(SizedTokenPool(Set.empty, currentLoans), None)
+  }
+
+  override def push(token: JobExecutionToken): SizedTokenPool = {
+    if (currentLoans.contains(token)) { SizedTokenPool(pool + token, currentLoans - token) } else this
+  }
+}
+
+final case class InfiniteTokenPool(tokenType: JobExecutionTokenType, override val currentLoans: Set[JobExecutionToken]) extends TokenPool {
```

Comment:
While I heartily applaud the belts & suspenders of the `final` here, why are you doing it here and not everywhere else? :)